sqld/src/replication/primary/logger.rs

```diff
@@ -720,28 +743,89 @@ pub struct ReplicationLogger {
             .create(true)
             .write(true)
             .read(true)
             .open(log_path)?;
+
         let max_log_frame_count = max_log_size * 1_000_000 / LogFile::FRAME_SIZE as u64;
         let log_file = LogFile::new(file, max_log_frame_count)?;
+        let header = log_file.header();
+
+        let should_recover = if header.version < 2 || header.sqld_version() != Version::current() {
+            tracing::info!("replication log version not compatible with current sqld version, recovering from database file.");
+            true
+        } else if fresh && data_path.exists() {
+            tracing::info!("replication log not found, recovering from database file.");
+            true
+        } else {
+            false
+        };
+
+        if should_recover {
+            Self::recover(log_file, data_path)
+        } else {
+            Self::from_log_file(db_path.to_path_buf(), log_file)
+        }
+    }
 
-        let header = log_file.header;
+    fn from_log_file(db_path: PathBuf, log_file: LogFile) -> anyhow::Result<Self> {
+        let header = log_file.header();
         let generation_start_frame_no = header.start_frame_no + header.frame_count;
 
         let (new_frame_notifier, _) = watch::channel(generation_start_frame_no);
 
         Ok(Self {
             generation: Generation::new(generation_start_frame_no),
-            compactor: LogCompactor::new(db_path, log_file.header.db_id)?,
+            compactor: LogCompactor::new(&db_path, log_file.header.db_id)?,
             log_file: RwLock::new(log_file),
-            db_path: db_path.to_owned(),
+            db_path,
             new_frame_notifier,
         })
     }
 
+    fn recover(log_file: LogFile, mut data_path: PathBuf) -> anyhow::Result<Self> {
+        tracing::info!("Replication log not found, or corrupted: recovering from database file");
+        let mut log_file = log_file.reset()?;
+
+        // It is necessary to checkpoint before we restore the replication log, since the WAL may
+        // contain pages that are not in the database file.
+        checkpoint_db(&data_path)?;
+        let snapshot_path = data_path.parent().unwrap().join("snapshots");
+        // best effort, there may be no snapshots
+        let _ = remove_dir_all(snapshot_path);
+
+        let data_file = File::open(&data_path)?;
+        let size = data_path.metadata()?.len();
+        ensure!(
+            size % WAL_PAGE_SIZE as u64 == 0,
+            "database file size is not a multiple of page size"
+        );
+        let num_page = size / WAL_PAGE_SIZE as u64;
+        let mut buf = [0; WAL_PAGE_SIZE as usize];
+        let mut page_no = 1; // page numbering starts at 1
+        for i in 0..num_page {
+            data_file.read_exact_at(&mut buf, i * WAL_PAGE_SIZE as u64)?;
+            log_file.push_page(&WalPage {
+                page_no,
+                size_after: dbg!((i == num_page - 1).then_some(num_page).unwrap_or_default()) as _,
```

Comment:
Looks like leftover `dbg!()` call?

Reply:
+ nitpick: I find `if i == num_page - 1 { num_page } else { 0 }` a little more human readable (: and potentially easier for the compiler to translate into simple ops